# Проект Определение стоимости автомобилей

## Введение


Сервис по продаже автомобилей с пробегом «Не бит, не крашен» разрабатывает приложение, чтобы привлечь новых клиентов. В нём можно будет узнать рыночную стоимость своего автомобиля. 
Постройте модель, которая умеет её определять. В вашем распоряжении данные о технических характеристиках, комплектации и ценах других автомобилей.
Критерии, которые важны заказчику:
- качество предсказания;
- время обучения модели;
- время предсказания модели.


In [40]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, MinMaxScaler, RobustScaler, LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.metrics import roc_auc_score, precision_score, make_scorer
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
from matplotlib.axes._axes import _log as matplotlib_axes_logger
from phik import phik_matrix
from phik.report import plot_correlation_matrix
from sklearn.metrics import mean_squared_error
import warnings
from sklearn.svm import SVC, SVR

matplotlib_axes_logger.setLevel('ERROR')
warnings.simplefilter(action='ignore', category=FutureWarning)
pd.set_option('max_colwidth', -1)


sns.set()

## Загрузка данных

In [41]:
df_cars = pd.read_csv('https://code.s3.yandex.net/datasets/autos.csv')
display(df_cars.head(5))
df_cars.info()

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
0  2016-03-24 11:52:17  480    NaN         1993              manual  0       
1  2016-03-24 10:58:45  18300  coupe       2011              manual  190     
2  2016-03-14 12:52:21  9800   suv         2004              auto    163     
3  2016-03-17 16:54:04  1500   small       2001              manual  75      
4  2016-03-31 17:25:20  3600   small       2008              manual  69      

   Model  Kilometer  RegistrationMonth  FuelType       Brand Repaired  \
0  golf   150000     0                  petrol    volkswagen  NaN       
1  NaN    125000     5                  gasoline  audi        yes       
2  grand  125000     8                  gasoline  jeep        NaN       
3  golf   150000     6                  petrol    volkswagen  no        
4  fabia  90000      7                  gasoline  skoda       no        

           DateCreated  NumberOfPictures  PostalCode             LastSeen  
0  2016-03-24 00:00:00  0                 70435       2016-04-07 03:16:57  
1  2016-03-24 00:00:00  0                 66954       2016-04-07 01:46:50  
2  2016-03-14 00:00:00  0                 90480       2016-04-05 12:47:46  
3  2016-03-17 00:00:00  0                 91074       2016-03-17 17:40:17  
4  2016-03-31 00:00:00  0                 60437       2016-04-06 10:17:21

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Kilometer          354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  Repaired           283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

По загруженным данным можно сделать следующие выводы:
- размерность датасета 354369 х 16
- датасет содержит как числовые так и категориальные типы данных. Кроме полей которые содержат дату и имеют тип object (DateCrawled, DateCreated, LastSeen), все остальные типы соответствуют своим данным. Типы колонок DateCrawled, DateCreated, LastSeen надо перобразовать в соответствующие типы (date или timestamp)
- в некоторых колонках имеются пропуски

## Предобработка данных

- Колонки DateCrawled, DateCreated, LastSeen приведены к типу datetime64
- Остальные колонки типа object приведены к типу string

In [42]:
df_cars_str_col_list = df_cars.select_dtypes(exclude=np.number).columns
df_cars[df_cars_str_col_list] = df_cars[df_cars_str_col_list].astype('string')

df_cars['DateCrawled'] = pd.to_datetime(df_cars['DateCrawled'])
df_cars['DateCreated'] = pd.to_datetime(df_cars['DateCreated'])
df_cars['LastSeen'] = pd.to_datetime(df_cars['LastSeen'])

df_cars.info()
df_cars.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   DateCrawled        354369 non-null  datetime64[ns]
 1   Price              354369 non-null  int64         
 2   VehicleType        316879 non-null  string        
 3   RegistrationYear   354369 non-null  int64         
 4   Gearbox            334536 non-null  string        
 5   Power              354369 non-null  int64         
 6   Model              334664 non-null  string        
 7   Kilometer          354369 non-null  int64         
 8   RegistrationMonth  354369 non-null  int64         
 9   FuelType           321474 non-null  string        
 10  Brand              354369 non-null  string        
 11  Repaired           283215 non-null  string        
 12  DateCreated        354369 non-null  datetime64[ns]
 13  NumberOfPictures   354369 non-null  int64   

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
0 2016-03-24 11:52:17  480    <NA>        1993              manual  0       
1 2016-03-24 10:58:45  18300  coupe       2011              manual  190     
2 2016-03-14 12:52:21  9800   suv         2004              auto    163     
3 2016-03-17 16:54:04  1500   small       2001              manual  75      
4 2016-03-31 17:25:20  3600   small       2008              manual  69      

   Model  Kilometer  RegistrationMonth  FuelType       Brand Repaired  \
0  golf   150000     0                  petrol    volkswagen  <NA>      
1  <NA>   125000     5                  gasoline  audi        yes       
2  grand  125000     8                  gasoline  jeep        <NA>      
3  golf   150000     6                  petrol    volkswagen  no        
4  fabia  90000      7                  gasoline  skoda       no        

  DateCreated  NumberOfPictures  PostalCode            LastSeen  
0 2016-03-24   0                 70435      2016-04-07 03:16:57  
1 2016-03-24   0                 66954      2016-04-07 01:46:50  
2 2016-03-14   0                 90480      2016-04-05 12:47:46  
3 2016-03-17   0                 91074      2016-03-17 17:40:17  
4 2016-03-31   0                 60437      2016-04-06 10:17:21

Колонка NumberOfPictures содержит значения все равные 0. В связи с этим она признана неинформативной и удалена из датасета

In [43]:
df_cars.describe()
df_cars = df_cars.drop(['NumberOfPictures'], axis = 1)
df_cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 15 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   DateCrawled        354369 non-null  datetime64[ns]
 1   Price              354369 non-null  int64         
 2   VehicleType        316879 non-null  string        
 3   RegistrationYear   354369 non-null  int64         
 4   Gearbox            334536 non-null  string        
 5   Power              354369 non-null  int64         
 6   Model              334664 non-null  string        
 7   Kilometer          354369 non-null  int64         
 8   RegistrationMonth  354369 non-null  int64         
 9   FuelType           321474 non-null  string        
 10  Brand              354369 non-null  string        
 11  Repaired           283215 non-null  string        
 12  DateCreated        354369 non-null  datetime64[ns]
 13  PostalCode         354369 non-null  int64   

Обнаружено 5 дубликатов для всех входных аттрибутов (исключая целевой аттрибут Price). Пока с ними ничего не буду делать, так как далее будет иметь место заполнение пропусков и я посмотрю на дубликаты после этой процедуры.

In [49]:
print(f'Количество дубликатов: df_cars: {df_cars.duplicated(subset=df_cars.columns.difference(["Price"])).sum()}')
df_cars[df_cars.duplicated(subset=df_cars.columns.difference(["Price"]))]

Количество дубликатов: df_cars: 5


DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
171088 2016-03-08 18:42:48  1799   coupe       1999              auto     
231258 2016-03-28 00:56:10  1000   small       2002              manual   
258109 2016-04-03 09:01:15  4699   coupe       2003              auto     
290453 2016-03-23 15:38:13  1890   sedan       1998              manual   
325651 2016-03-18 18:46:15  1999   wagon       2001              manual   

        Power   Model  Kilometer  RegistrationMonth  FuelType          Brand  \
171088  193    clk     20000      7                  petrol    mercedes_benz   
231258  83     other   150000     1                  petrol    suzuki          
258109  218    clk     125000     6                  petrol    mercedes_benz   
290453  75     golf    125000     12                 petrol    volkswagen      
325651  131    passat  150000     7                  gasoline  volkswagen      

       Repaired DateCreated  PostalCode            LastSeen  
171088  no      2016-03-08   89518      2016-03-09 09:46:57  
231258  no      2016-03-28   66589      2016-03-28 08:46:21  
258109  yes     2016-04-03   75196      2016-04-07 09:44:54  
290453  no      2016-03-23   58089      2016-03-24 23:15:33  
325651  no      2016-03-18   36391      2016-03-18 18:46:15

Ананлиз данных показал, что пропуски имеют место только в категориальных колонках (кроме колонки Brand) и в связи с этим принято решение заполнить их в пайплайне модой для соответствуующего аттрибута. После этого еще раз провести проверку на дубликаты (без целевого признака Price) и уже после этого удалить имеющиеся дубликаты.

In [57]:
print("Cat features")

for c in ['VehicleType', 'Gearbox', 'Model', 'FuelType', 'Brand', 'Repaired']:
    print(f"------- {c} -------")
    display(df_cars[c].value_counts(dropna = False))

df_cars.info()

Cat features
------- VehicleType -------


sedan          91457
small          79831
wagon          65166
NaN            37490
bus            28775
convertible    20203
coupe          16163
suv            11996
other          3288 
Name: VehicleType, dtype: Int64

------- Gearbox -------


manual    268251
auto      66285 
NaN       19833 
Name: Gearbox, dtype: Int64

------- Model -------


golf                  29232
other                 24421
3er                   19761
NaN                   19705
polo                  13066
                      ...  
i3                    8    
rangerover            4    
serie_3               4    
serie_1               2    
range_rover_evoque    2    
Name: Model, Length: 251, dtype: Int64

------- FuelType -------


petrol      216352
gasoline    98720 
NaN         32895 
lpg         5310  
cng         565   
hybrid      233   
other       204   
electric    90    
Name: FuelType, dtype: Int64

------- Brand -------


volkswagen        77013
opel              39931
bmw               36914
mercedes_benz     32046
audi              29456
ford              25179
renault           17927
peugeot           10998
fiat              9643 
seat              6907 
mazda             5615 
skoda             5500 
smart             5246 
citroen           5148 
nissan            4941 
toyota            4606 
hyundai           3587 
sonstige_autos    3374 
volvo             3210 
mini              3202 
mitsubishi        3022 
honda             2817 
kia               2465 
suzuki            2323 
alfa_romeo        2314 
chevrolet         1754 
chrysler          1439 
dacia             900  
daihatsu          806  
subaru            762  
porsche           758  
jeep              677  
trabant           589  
land_rover        545  
daewoo            542  
saab              526  
jaguar            505  
rover             486  
lancia            471  
lada              225  
Name: Brand, dtype: Int64

------- Repaired -------


no     247161
NaN    71154 
yes    36054 
Name: Repaired, dtype: Int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 15 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   DateCrawled        354369 non-null  datetime64[ns]
 1   Price              354369 non-null  int64         
 2   VehicleType        316879 non-null  string        
 3   RegistrationYear   354369 non-null  int64         
 4   Gearbox            334536 non-null  string        
 5   Power              354369 non-null  int64         
 6   Model              334664 non-null  string        
 7   Kilometer          354369 non-null  int64         
 8   RegistrationMonth  354369 non-null  int64         
 9   FuelType           321474 non-null  string        
 10  Brand              354369 non-null  string        
 11  Repaired           283215 non-null  string        
 12  DateCreated        354369 non-null  datetime64[ns]
 13  PostalCode         354369 non-null  int64   

## Исследовательский анализ данных